### ISSUE TITLE
[Question]TFLite gpu acceleration.

### ISSUE THREAD

**Issue by nanamare** [tensorflow/tensorflow#18199]
Current, my model's inference speed (mate 10) 300ms
We have applied all the structural transformations that can be done in the model, algorithm
So, I am looking for a way to improve the performance of the Android platform.
Example of Render Script, OpenCL...
But, don't know if these ways really improve inference performance dramatically.
I wonder if these methods are applied to Tensorflow Mobile(including TFLite) at present.
I also wonder which method is the most appropriate.

Have a nice day :)

**Comment by VincentLin78** [comment]
There's standard channel for hw acceleration called NNAPI in Android 8.1.
https://developer.android.com/ndk/guides/neuralnetworks/index.html

TFL is gonna built on top of it and dispatch model to it.
Depends on how SOC vendor implemented NN HAL, the dispatched model may or may not have performance gain.

Before doing that, make sure if your model can be converted to tflite format by TOCO.
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/lite/toco
If yes, then you can run your converted tflite model with TFL.

**Comment by nanamare** [comment]
@VincentLin78 
I try to convert pb to tflite, but encountered unsupported operation error ([here is my posting](https://github.com/tensorflow/tensorflow/issues/18191)).  I converted 5 pb  to tflite, but there was no dramatic effect. (mobilenet, inception, ssd mobilenet etc)
also
[https://developer.android.com/ndk/guides/neuralnetworks/index.html]
I'm wondering if you need to write code that references the model(load graph, open session), or additionally needs to add other parts(model operation add, mul, activation function etc).

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 75 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by llhe** [comment]
@nanamare 
To run neural nets on mobile phones with GPU support, you can checkout [MACE](https://github.com/xiaomi/mace), which supports Adreno, Mali and PowerVR GPUs. Here are some [benchmark results](https://github.com/XiaoMi/mace/issues/1).

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by aselle** [comment]
@nanamare, we are working on various acceleration strategies (including GPU), and we will update this post when we can.

Thanks!

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shashishekhar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by oscarriddle** [comment]
@aselle @shashishekhar 
Hi, I just wondering whether tflite will support CUDA gpu acceleration?
I think tflite provides a more compatible quantization over difference types of nodes in tf.

Thanks,

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shashishekhar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by shashishekhar** [comment]
@oscarriddle : CUDA may not be available on many mobile devices.  
In short term TFLite will support acceleration using OpenGL and other drivers that can target GPU. 

Technically, it is certainly possible to write a [delegate](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/lite/delegates) that uses CUDA for acceleration.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shashishekhar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shashishekhar: It has been 32 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shashishekhar: It has been 46 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by wosiu** [comment]
@aselle @shashishekhar Is it already known when acceleration using OpenGL is about to happen?
Just FYI:
I run a model on a different Android devices (Android < 8.0) and interference time takes 400-700ms using tflite. The same model on iOS takes ~30ms using CoreML.

**Comment by achowdhery** [comment]
In short term TFLite will soon support acceleration using OpenGL and other drivers that can target GPU.

**Comment by legion-zver** [comment]
Still would Firebase Mlkit supported the acceleration out of the box, it would be cool... But as long as it's scary!

**Comment by idstein** [comment]
We are looking forward to the GPU OpenGLES delegate. But out benchmarks show that on same device Mace OpenCL is at least 30% faster than TfLite NNAPI with OpenGL NN driver.

Any news of alternative delegate drivers?

Will the delegate source be made public soon? We would like to implement additional ops such as resize nearest neighbor.

**Comment by miaout17** [comment]
The binary preview of GPU OpenGLES delegate backend was just released. See also the [blog post](https://medium.com/tensorflow/tensorflow-lite-now-faster-with-mobile-gpus-developer-preview-e15797e6dee7?linkId=62443226) and [tutorial document](https://www.tensorflow.org/lite/performance/gpu)

**Comment by miaout17** [comment]
Since we released the GPU backend binary. I'm closing this issue. 
Feel free to create other issues if you have more specific questions.

**Comment by noah003** [comment]
> Since we released the GPU backend binary. I'm closing this issue.
> Feel free to create other issues if you have more specific questions.

Is the source code of the GPU delegate has been open source? I can not find the source code, just find the "libtensorflowlite_gpu_jni.so"

**Comment by patrick-ucr** [comment]
I am working on TFLite using Mobile GPU. Just a followup question, "does TFLite support mobile GPU for detector task (instead of classifier in the demo)?" Also, "does TFLite's mobile GPU only works with floating-point models not quantized ones?" (I found this from code comments)
